tmpl/wiki/mirage2-to-mirage3.md

```diff
@@ -0,0 +1,1416 @@
+
+```
+~/mirage-skeleton/stackv4$ make
+mirage build
+ocamlfind ocamldep -package tcpip.udp -package tcpip.tcp -package tcpip.stack-direct -package tcpip.ipv4 -package tcpip.icmpv4 -package tcpip.ethif -package tcpip.arpv4 -package tcpip -package mirage-unix -package mirage-types-lwt -package mirage-types -package mirage-runtime -package mirage-random -package mirage-net-unix -package mirage-logs -package mirage-console-unix -package mirage-clock-unix -package lwt -package functoria-runtime -predicates mirage_unix -modules unikernel.ml > unikernel.ml.depends
+ocamlfind ocamlc -c -g -g -bin-annot -safe-string -principal -strict-sequence -package tcpip.udp -package tcpip.tcp -package tcpip.stack-direct -package tcpip.ipv4 -package tcpip.icmpv4 -package tcpip.ethif -package tcpip.arpv4 -package tcpip -package mirage-unix -package mirage-types-lwt -package mirage-types -package mirage-runtime -package mirage-random -package mirage-net-unix -package mirage-logs -package mirage-console-unix -package mirage-clock-unix -package lwt -package functoria-runtime -predicates mirage_unix -w A-4-41-42-44 -color always -o unikernel.cmo unikernel.ml
++ ocamlfind ocamlc -c -g -g -bin-annot -safe-string -principal -strict-sequence -package tcpip.udp -package tcpip.tcp -package tcpip.stack-direct -package tcpip.ipv4 -package tcpip.icmpv4 -package tcpip.ethif -package tcpip.arpv4 -package tcpip -package mirage-unix -package mirage-types-lwt -package mirage-types -package mirage-runtime -package mirage-random -package mirage-net-unix -package mirage-logs -package mirage-console-unix -package mirage-clock-unix -package lwt -package functoria-runtime -predicates mirage_unix -w A-4-41-42-44 -color always -o unikernel.cmo unikernel.ml
+File "unikernel.ml", line 17, characters 4-11:
+Error: Unbound value C.log_s
+Hint: Did you mean log?
+Command exited with code 2.
+run ['ocamlbuild' '-use-ocamlfind' '-classic-display' '-tags'
+     'predicate(mirage_unix),warn(A-4-41-42-44),color(always),debug,bin_annot,strict_sequence,principal,safe_string'
+     '-pkgs'
+     'functoria-runtime,lwt,mirage-clock-unix,mirage-console-unix,mirage-logs,mirage-net-unix,mirage-random,mirage-runtime,mirage-types,mirage-types-lwt,mirage-unix,tcpip,tcpip.arpv4,tcpip.ethif,tcpip.icmpv4,tcpip.ipv4,tcpip.stack-direct,tcpip.tcp,tcpip.udp'
+     '-cflags' '-g' '-lflags' '-g' '-tag-line' '<static*.*>: warn(-32-34)'
+     '-X' '_build-ukvm' 'main.native']: exited with 10
+Makefile:18: recipe for target 'build' failed
+make: *** [build] Error 1
+```
+
+### name changes: C.log_s -> C.log, T.get_dest -> T.dst
+
+`C` is a module of type `Mirage_console_lwt.S`.  This module type was refactored in Mirage 3 to remove `log`, and rename the more commonly used `log_s` to `log`.  If we really want to be sure that our messages go to the *console*, rather than to another available logging destination, we can change `unikernel.ml` to use `C.log` instead of `C.log_s`:
+
```

Comment:
Might be worth adding a hint here that we'll talk about logs lower down, before people spend time converting their logging code to the new console interface. People might get here by googling the error and not read the whole thing.